nabl2.terms/src/main/java/mb/nabl2/regexp/RegExpMatcher.java

```diff
@@ -136,4 +137,20 @@ private RegExpMatcher(IRegExp<S> state, Map<IRegExp<S>, Map<S, IRegExp<S>>> stat
         return new RegExpMatcher<>(initial, stateTransitions, defaultTransitions, nonFinal, isNullable);
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        RegExpMatcher<?> that = (RegExpMatcher<?>) o;
+        return Objects.equals(state, that.state) &&
```

Comment:
It is enough to only compare `state` here. All other fields are derived from that regular expression.